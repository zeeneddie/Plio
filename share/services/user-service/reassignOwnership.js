import { reduce, lensPath, lensProp, set, when, evolve, equals, always } from 'ramda';

const improvementPlan = ['improvementPlan', 'owner'];
const analysis = [['analysis', 'executor'], ['analysis', 'completedBy']];
const updateOfStandards = [
  ['updateOfStandards', 'executor'],
  ['updateOfStandards', 'completedBy'],
];
const problemFields = [
  'ownerId',
  'originatorId',
  improvementPlan,
  ...analysis,
  ...updateOfStandards,
];
const canvasFields = ['originatorId'];

export default async function reassignOwnership(args, context) {
  const { organizationId, userId, ownerId } = args;
  const {
    collections: {
      Actions,
      LessonsLearned,
      NonConformities,
      Risks,
      Standards,
      WorkItems,
      Goals,
      KeyPartners,
      KeyActivities,
      KeyResources,
      CostLines,
      CustomerRelationships,
      Channels,
      ValuePropositions,
      CustomerSegments,
      RevenueStreams,
      Organizations,
    },
  } = context;
  const applyReassign = (keys, collection, query = { organizationId }) => {
    const rawCollection = collection.rawCollection();
    return collection.find(query).map(async (doc) => {
      const transformations = reduce((acc, key) => set(
        Array.isArray(key) ? lensPath(key) : lensProp(key),
        when(equals(userId), always(ownerId)),
        acc,
      ), {}, keys);
      const transformed = evolve(transformations, doc);

      return rawCollection.save(transformed);
    });
  };

  const promises = [
    applyReassign(
      ['completedBy', 'toBeVerifiedBy', 'verifiedBy', 'ownerId', 'toBeCompletedBy'],
      Actions,
    ),
    applyReassign(['owner'], LessonsLearned),
    applyReassign(problemFields, NonConformities),
    applyReassign(problemFields, Risks),
    applyReassign(['owner', improvementPlan], Standards),
    applyReassign(['assigneeId'], WorkItems),
    applyReassign(['ownerId', 'completedBy'], Goals),
    applyReassign(canvasFields, KeyPartners),
    applyReassign(canvasFields, KeyActivities),
    applyReassign(canvasFields, KeyResources),
    applyReassign(canvasFields, CostLines),
    applyReassign(canvasFields, CustomerRelationships),
    applyReassign(canvasFields, Channels),
    applyReassign(canvasFields, ValuePropositions),
    applyReassign(canvasFields, CustomerSegments),
    applyReassign(canvasFields, RevenueStreams),
    applyReassign(
      [
        ['review', 'risks', 'reviewerId'],
        ['review', 'standards', 'reviewerId'],
      ],
      Organizations,
      { _id: organizationId },
    ),
  ];

  return Promise.all(promises).then(array => Promise.all(...array));
}
