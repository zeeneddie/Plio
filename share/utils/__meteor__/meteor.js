const { Mongo } = require('./mongo');

const Meteor = {
  isServer: true,
  loginWithPassword: jest.fn(),
  loginWithFacebook: jest.fn(),
  methods: jest.fn(),
  call: jest.fn(),
  publish: jest.fn(),
  subscribe: jest.fn(),
  user: jest.fn(),
  users: new Mongo.Collection(),
  userId: jest.fn().mockReturnValue('f00bar'),
  startup: jest.fn(),
  bindEnvironment: jest.fn(),
  wrapAsync: jest.fn(),
  Error: jest.fn(Error),
  absoluteUrl: jest.fn(url => url),
};

module.exports = { Meteor };
