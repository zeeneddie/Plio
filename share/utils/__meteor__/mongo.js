import MongoDBMemoryServer from 'mongodb-memory-server';
import { MongoClient } from 'mongodb';
import faker from 'faker';

let client;
let db;
let mongoServer;

export const __setupDB = jest.fn(async () => {
  mongoServer = new MongoDBMemoryServer();
  const uri = await mongoServer.getConnectionString();
  client = await MongoClient.connect(uri);
  db = client.db('test');
});

export const __closeDB = jest.fn(async () => {
  await client.close();
  await mongoServer.stop();
});

export const __clearDB = jest.fn(async () => {
  const collections = await db.collections();
  return Promise.all(collections.map(collection => collection.drop()));
});

const Collection = jest.fn(function (name) {
  if (!db) return null;

  this.collection = db.collection(name);
  this.rawCollection = () => this.collection;

  this.insert = jest.fn(async function insert(doc, ...args) {
    const writeDoc = { _id: faker.random.uuid(), ...doc };
    const { insertedId } = await this.collection.insertOne(writeDoc, ...args);
    return insertedId;
  });

  this.update = jest.fn(async function update(...args) {
    const { result: { nModified } } = await this.collection.update(...args);
    return nModified;
  });

  this.remove = jest.fn(async function remove(...args) {
    const { deletedCount } = await this.collection.deleteMany(...args);
    return deletedCount;
  });

  this.findOne = jest.fn(async function findOne(...args) {
    return this.collection.findOne(...args);
  });

  this.find = jest.fn(function find(...args) {
    const cursor = this.collection.find(...args);
    return {
      ...cursor,
      ...cursor.__proto__, // eslint-disable-line no-proto
      fetch: jest.fn(async () => cursor.toArray()),
      map: jest.fn(async (...mapArgs) => cursor.map(...mapArgs).toArray()),
    };
  });

  this.before = {
    insert: jest.fn(),
    update: jest.fn(),
  };

  this.after = {
    insert: jest.fn(),
    update: jest.fn(),
  };

  return this;
});

Collection.prototype.attachSchema = jest.fn();
Collection.prototype.helpers = jest.fn();

export const Mongo = { Collection };

const RemoteCollectionDriver = jest.fn();
RemoteCollectionDriver.prototype.open = jest.fn().mockReturnThis();
RemoteCollectionDriver.prototype.insert = jest.fn();
RemoteCollectionDriver.prototype.update = jest.fn();
RemoteCollectionDriver.prototype.remove = jest.fn();
RemoteCollectionDriver.prototype.findOne = jest.fn();
RemoteCollectionDriver.prototype.find = jest.fn(() => ({
  count: jest.fn(),
  fetch: jest.fn(),
}));
export const MongoInternals = { RemoteCollectionDriver };
