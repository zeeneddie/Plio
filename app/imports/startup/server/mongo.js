import { Mongo } from 'meteor/mongo';
import { Meteor } from 'meteor/meteor';

Meteor.startup(() => {
  Mongo.setConnectionOptions({
    mongos: {
      socketOptions: {
        socketTimeoutMS: 360000,
        keepAlive: true,
        reconnectTries: Number.MAX_VALUE,
      },
    },
  });
});
