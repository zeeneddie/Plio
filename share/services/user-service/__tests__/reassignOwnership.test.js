import { __setupDB, __closeDB, __clearDB } from 'meteor/mongo';
import faker from 'faker';

import createContext from '../../../utils/tests/createContext';
import reassignOwnership from '../reassignOwnership';

describe('UserService.reassignOwnership', () => {
  faker.seed(123);

  let context;
  let args;
  const organizationId = faker.random.uuid();
  const userId = faker.random.uuid();

  beforeAll(async () => {
    await __setupDB();

    context = createContext({});
    args = { organizationId, userId: context.userId, ownerId: userId };
  });
  afterAll(__closeDB);
  beforeEach(__clearDB);

  it('reassigns ownership of one document', async () => {
    const _id = await context.collections.Actions.insert({
      organizationId,
      ownerId: context.userId,
    });

    await reassignOwnership(args, context);

    await expect(context.collections.Actions.findOne({ _id })).resolves.toMatchObject({
      ownerId: userId,
    });
  });

  it('reassigns ownership of multiple documents', async () => {
    await context.collections.Actions.insert({ organizationId, ownerId: context.userId });
    await context.collections.Actions.insert({ organizationId, ownerId: context.userId });

    await reassignOwnership(args, context);

    const cursor = context.collections.Actions.find({ ownerId: userId });
    const count = await cursor.count();

    expect(count).toBe(2);
  });

  it('works with nested objects', async () => {
    const _id = await context.collections.NonConformities.insert({
      organizationId,
      improvementPlan: {
        owner: context.userId,
      },
    });

    await reassignOwnership(args, context);

    await expect(context.collections.NonConformities.findOne({ _id })).resolves.toMatchObject({
      improvementPlan: {
        owner: userId,
      },
    });
  });

  it('matches snapshot', async () => {
    const {
      Actions,
      LessonsLearned,
      NonConformities,
      Risks,
      Standards,
      WorkItems,
      Goals,
      KeyPartners,
      KeyActivities,
      KeyResources,
      CostLines,
      CustomerRelationships,
      Channels,
      ValuePropositions,
      CustomerSegments,
      RevenueStreams,
      Organizations,
    } = context.collections;

    await Promise.all([
      Actions.insert({
        organizationId,
        toBeCompletedBy: context.userId,
        completedBy: context.userId,
        toBeVerifiedBy: context.userId,
        verifiedBy: context.userId,
        ownerId: context.userId,
      }),
      LessonsLearned.insert({
        organizationId,
        owner: context.userId,
      }),
      NonConformities.insert({
        organizationId,
        ownerId: context.userId,
        originatorId: context.userId,
        improvementPlan: {
          owner: context.userId,
        },
        analysis: {
          executor: context.userId,
          completedBy: context.userId,
        },
        updateOfStandards: {
          executor: context.userId,
          completedBy: context.userId,
        },
      }),
      Risks.insert({
        organizationId,
        ownerId: context.userId,
        originatorId: context.userId,
        improvementPlan: {
          owner: context.userId,
        },
        analysis: {
          executor: context.userId,
          completedBy: context.userId,
        },
        updateOfStandards: {
          executor: context.userId,
          completedBy: context.userId,
        },
      }),
      Standards.insert({
        organizationId,
        owner: context.userId,
        improvementPlan: {
          owner: context.userId,
        },
      }),
      WorkItems.insert({
        organizationId,
        assigneeId: context.userId,
      }),
      Goals.insert({
        organizationId,
        ownerId: context.userId,
        completedBy: context.userId,
      }),
      KeyPartners.insert({
        organizationId,
        originatorId: context.userId,
      }),
      KeyActivities.insert({
        organizationId,
        originatorId: context.userId,
      }),
      KeyResources.insert({
        organizationId,
        originatorId: context.userId,
      }),
      CostLines.insert({
        organizationId,
        originatorId: context.userId,
      }),
      CustomerRelationships.insert({
        organizationId,
        originatorId: context.userId,
      }),
      Channels.insert({
        organizationId,
        originatorId: context.userId,
      }),
      ValuePropositions.insert({
        organizationId,
        originatorId: context.userId,
      }),
      CustomerSegments.insert({
        organizationId,
        originatorId: context.userId,
      }),
      RevenueStreams.insert({
        organizationId,
        originatorId: context.userId,
      }),
      Organizations.insert({
        _id: organizationId,
        review: {
          risks: {
            reviewerId: context.userId,
          },
          standards: {
            reviewerId: context.userId,
          },
        },
      }),
    ]);

    await reassignOwnership(args, context);

    const promises = [
      Actions,
      LessonsLearned,
      NonConformities,
      Risks,
      Standards,
      WorkItems,
      Goals,
      KeyPartners,
      KeyActivities,
      KeyResources,
      CostLines,
      CustomerRelationships,
      Channels,
      ValuePropositions,
      CustomerSegments,
      RevenueStreams,
    ]
      .map(collection => collection.find({ organizationId }).fetch())
      .concat(Organizations.find({ _id: organizationId }).fetch());
    const results = await Promise.all(promises);

    expect(results).toMatchSnapshot();
  });
});
