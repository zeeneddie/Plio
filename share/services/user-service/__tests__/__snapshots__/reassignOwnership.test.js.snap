// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService.reassignOwnership matches snapshot 1`] = `
Array [
  Array [
    Object {
      "_id": "86a5c5d2-6d05-4788-9984-0efae5384784",
      "completedBy": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "ownerId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "toBeCompletedBy": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "toBeVerifiedBy": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "verifiedBy": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "e4f94b98-c56a-4bd2-a9fd-5fd11603e7e8",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "owner": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "3104de39-071c-47b8-86b4-d62ccc4a4fa6",
      "analysis": Object {
        "completedBy": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
        "executor": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      },
      "improvementPlan": Object {
        "owner": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      },
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "ownerId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "updateOfStandards": Object {
        "completedBy": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
        "executor": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      },
    },
  ],
  Array [
    Object {
      "_id": "13a7e8b6-7270-4c99-81e9-9d752e0c295c",
      "analysis": Object {
        "completedBy": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
        "executor": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      },
      "improvementPlan": Object {
        "owner": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      },
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "ownerId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "updateOfStandards": Object {
        "completedBy": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
        "executor": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      },
    },
  ],
  Array [
    Object {
      "_id": "71ad3d9e-e211-472b-a16d-861737c57ecd",
      "improvementPlan": Object {
        "owner": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      },
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "owner": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "0a87e996-34b4-46ba-909a-70ab67b1f811",
      "assigneeId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
    },
  ],
  Array [
    Object {
      "_id": "0a9fcc89-e14b-47af-85c3-8465ca607c29",
      "completedBy": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "ownerId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "5adb2861-f732-4eae-81a7-5875d0231173",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "0c4e6054-5fd8-48a8-817c-f6611278f755",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "c1c93379-b46e-439f-baad-f0e408619f23",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "9553f193-a46a-4b46-9c0f-5287289644a6",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "0dea0f27-ce0c-4b4f-8ddb-5ff6adb57e12",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "ffad1d21-d940-48b1-8915-35ea558b9a3f",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "94a3416e-70e7-490f-a858-a057eeb310b6",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "e58df07e-0a8e-4899-8d55-5e66c65631f7",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "ea54ae47-3456-456e-8dbb-7f86d3b184b2",
      "organizationId": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "originatorId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
    },
  ],
  Array [
    Object {
      "_id": "bb463b8b-b76c-4f6a-9726-65ab5730b69b",
      "review": Object {
        "risks": Object {
          "reviewerId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
        },
        "standards": Object {
          "reviewerId": "a27218b8-6a4d-47bb-95b6-5a55334fac1c",
        },
      },
    },
  ],
]
`;
